Adding pkgsStatic: a fully static overlay

matthewbauer opened the pull request:
###### Motivation for this change

This uses the cross compilation infrastructure to build a statically linked Musl-based package set. Some basic things seem to build okay... Working on getting things working as much as possible.

You can access them from the pkgsStatic attribute set. Basic things like pkgsStatic.hello seem to work okay.

As part of the implementation this adds a new type of overlay called "crossOverlays". These are similar to ordinary overlays except for one critical difference. They only apply to the /target/ package set. To accomplish this, we use the cross stdenv to clearly separate out build, host, and target.

In addition, I have added a custom "static" overlay that will enable static by modifying the stdenv and modifying other packages as necessary. Additional packages can be enabled later on here.

/cc @nh2 @dtzWill

Ericson2314 commented:
https://github.com/NixOS/nixpkgs/pull/48286 might help you

nh2 commented:
@matthewbauer @dtzWill Are any of you at NixCon by any chance?

dtzWill commented:
No :(.  Watching the stream, though!

On Thu, 25 Oct 2018 13:14:05 +0000 (UTC), Niklas Hambüchen <notifications@github.com> wrote:
> @matthewbauer @dtzWill Are any of you at NixCon by any chance?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixpkgs/pull/48803#issuecomment-433045406Non-text part: text/html

matthewbauer commented:
Nope but @ericson2314 should be

Ericson2314 commented:
Actually I won't be :(

Ericson2314 commented:
Maybe next year I'll finally learn not to procrastinate.

Ericson2314 commented on pkgs/development/libraries/ncurses/default.nix:
```diff
@@ -28,7 +29,7 @@ stdenv.mkDerivation rec {
 
   configureFlags = [
-    "--with-shared"
+    (if enableShared then "--with-shared" else "--without-shared")
```
Can use `lib.withFeature` for this

GrahamcOfBorg commented:
<!--REQUEST_ID=9f1c7511-fc86-4c20-992f-b6a86c8e9a7d-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=4ee853a3-887c-4b77-869e-fe1acaa33f4b)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/wydzkhfm25cgbr1nfq28nqvgk0mk7518-ncurses-6.1
/nix/store/x0aj2fsd3hskn1fckk3w4b8bl6mx8m1s-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=9f1c7511-fc86-4c20-992f-b6a86c8e9a7d-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=c221de8e-c062-4e8f-8898-ab752e216ec0)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/snmcl9nfjj49h2j0s1qpvmsajhm91141-ncurses-6.1
/nix/store/v8w9k5zzfidw4ik5rjxbpf9cfvbrj9s3-zlib-1.2.11
```
</p></details>

matthewbauer commented:
@grahamcofborg build pkgsStatic.hello

GrahamcOfBorg commented:
<!--REQUEST_ID=9f1c7511-fc86-4c20-992f-b6a86c8e9a7d-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=fb23fd31-9c8c-4ec7-9dc8-756acd65232a)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/iirmif7qgp7pgbv80z5x1sj3hbay893d-ncurses-6.1
/nix/store/di0d17idlw6fxf3qwpfyxciw94ywi664-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=0ccb3da1-5b33-49d4-8f69-249434d10552-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=056cbbd8-6ec9-4d2f-9c13-6cd5812f8505)

Attempted: pkgsStatic.hello

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/2zxr3jgljngkcmv3dkxcja3698pjrx1s-hello-2.10-aarch64-unknown-linux-musl
shrinking /nix/store/2zxr3jgljngkcmv3dkxcja3698pjrx1s-hello-2.10-aarch64-unknown-linux-musl/bin/hello
cannot find section .dynamic
gzipping man pages under /nix/store/2zxr3jgljngkcmv3dkxcja3698pjrx1s-hello-2.10-aarch64-unknown-linux-musl/share/man/
aarch64-unknown-linux-musl-strip is /nix/store/bqvzvb5b9c7rl4k5zara50gwnm5mmasr-aarch64-unknown-linux-musl-binutils-2.30/bin/aarch64-unknown-linux-musl-strip
stripping (with command aarch64-unknown-linux-musl-strip and flags -S) in /nix/store/2zxr3jgljngkcmv3dkxcja3698pjrx1s-hello-2.10-aarch64-unknown-linux-musl/lib  /nix/store/2zxr3jgljngkcmv3dkxcja3698pjrx1s-hello-2.10-aarch64-unknown-linux-musl/bin
patching script interpreter paths in /nix/store/2zxr3jgljngkcmv3dkxcja3698pjrx1s-hello-2.10-aarch64-unknown-linux-musl
checking for references to /build in /nix/store/2zxr3jgljngkcmv3dkxcja3698pjrx1s-hello-2.10-aarch64-unknown-linux-musl...
cannot find section .dynamic
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=0ccb3da1-5b33-49d4-8f69-249434d10552-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=468103dd-88e6-4c75-b0cc-45d4f5b30978)

Attempted: pkgsStatic.hello

<details><summary>Partial log (click to expand)</summary><p>

```
 /nix/store/acnfbicd84bnya2l2dq868b5l482qihw-coreutils-8.30/bin/mkdir -p '/nix/store/fvdzx3bmx2mxcavg4m6lja21h4r6bv09-hello-2.10/share/man/man1'
 /nix/store/acnfbicd84bnya2l2dq868b5l482qihw-coreutils-8.30/bin/install -c -m 644 hello.1 '/nix/store/fvdzx3bmx2mxcavg4m6lja21h4r6bv09-hello-2.10/share/man/man1'
make[3]: Leaving directory '/private/tmp/nix-build-hello-2.10.drv-0/hello-2.10'
make[2]: Leaving directory '/private/tmp/nix-build-hello-2.10.drv-0/hello-2.10'
make[1]: Leaving directory '/private/tmp/nix-build-hello-2.10.drv-0/hello-2.10'
post-installation fixup
gzipping man pages under /nix/store/fvdzx3bmx2mxcavg4m6lja21h4r6bv09-hello-2.10/share/man/
strip is /nix/store/g5r4apl0za012ffs6ladinwa5w0m1l3k-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/fvdzx3bmx2mxcavg4m6lja21h4r6bv09-hello-2.10/bin
patching script interpreter paths in /nix/store/fvdzx3bmx2mxcavg4m6lja21h4r6bv09-hello-2.10
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=023fafdb-c56d-4a5b-8292-c3de6bc9927c-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=33777719-a638-436f-95a1-55314a582167)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/iirmif7qgp7pgbv80z5x1sj3hbay893d-ncurses-6.1
/nix/store/di0d17idlw6fxf3qwpfyxciw94ywi664-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=023fafdb-c56d-4a5b-8292-c3de6bc9927c-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=74468380-d12f-486d-bbc4-586f03a4070d)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/snmcl9nfjj49h2j0s1qpvmsajhm91141-ncurses-6.1
/nix/store/v8w9k5zzfidw4ik5rjxbpf9cfvbrj9s3-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=023fafdb-c56d-4a5b-8292-c3de6bc9927c-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=4eccfadb-0c2e-4afa-ad0e-9b4d146e7d50)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/wydzkhfm25cgbr1nfq28nqvgk0mk7518-ncurses-6.1
/nix/store/x0aj2fsd3hskn1fckk3w4b8bl6mx8m1s-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=94c86f28-883f-4a38-a0ee-d8fcfa033d9f-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=46fd4895-015d-4823-b7d0-7276f8c01604)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/wydzkhfm25cgbr1nfq28nqvgk0mk7518-ncurses-6.1
/nix/store/x0aj2fsd3hskn1fckk3w4b8bl6mx8m1s-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=94c86f28-883f-4a38-a0ee-d8fcfa033d9f-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=dbaef6cc-0a88-4348-86ba-e463fc125d4c)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/snmcl9nfjj49h2j0s1qpvmsajhm91141-ncurses-6.1
/nix/store/v8w9k5zzfidw4ik5rjxbpf9cfvbrj9s3-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=94c86f28-883f-4a38-a0ee-d8fcfa033d9f-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=8a40717d-f7a7-4bc5-adeb-8d8b1689cab4)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/iirmif7qgp7pgbv80z5x1sj3hbay893d-ncurses-6.1
/nix/store/di0d17idlw6fxf3qwpfyxciw94ywi664-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=98a08d3e-9af9-408c-849f-c9463fa537fe-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=2d0d4a16-66c3-44af-8cb5-cd649d839b37)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/wydzkhfm25cgbr1nfq28nqvgk0mk7518-ncurses-6.1
/nix/store/x0aj2fsd3hskn1fckk3w4b8bl6mx8m1s-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=98a08d3e-9af9-408c-849f-c9463fa537fe-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=b61337f7-81af-44a5-bf4e-b8cdd442aa76)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/iirmif7qgp7pgbv80z5x1sj3hbay893d-ncurses-6.1
/nix/store/di0d17idlw6fxf3qwpfyxciw94ywi664-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=98a08d3e-9af9-408c-849f-c9463fa537fe-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=93c514bc-709f-46b2-bea7-8419d4279b61)

Attempted: ncurses, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/snmcl9nfjj49h2j0s1qpvmsajhm91141-ncurses-6.1
/nix/store/v8w9k5zzfidw4ik5rjxbpf9cfvbrj9s3-zlib-1.2.11
```
</p></details>

matthewbauer committed:
c-ares: don’t set configure flags

These aren’t needed with dontDisableStatic

GrahamcOfBorg commented:
<!--REQUEST_ID=87c2a285-87cc-46b3-8432-233242f676de-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=9cdc4a66-92a7-47f7-811b-4383fedcba48)

Attempted: c-ares, libxml2, ncurses, openssl, perl, zlib

The following builds were skipped because they don't evaluate on x86_64-darwin: proot

<details><summary>Partial log (click to expand)</summary><p>

```
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/v0yghpanchlkix2xjd6gci3w3hppmx8a-c-ares-1.14.0
/nix/store/fy1lmal54cgkr714fv1iy3m5wprjks2h-libxml2-2.9.8-bin
/nix/store/snmcl9nfjj49h2j0s1qpvmsajhm91141-ncurses-6.1
/nix/store/89i7jmd8kc5lc4da0xmd2lbddzzmrnpg-openssl-1.0.2p-bin
/nix/store/rnbicw8zk4273agv4vla8kqbxpyqf8n9-perl-5.28.0
/nix/store/v8w9k5zzfidw4ik5rjxbpf9cfvbrj9s3-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=87c2a285-87cc-46b3-8432-233242f676de-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=9537b3a8-61e0-47d7-8242-ea7a59f50ecf)

Attempted: c-ares, libxml2, ncurses, openssl, perl, proot, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/0b5bvyqz6hy754c5cm2imsln59kjprvd-proot-5.1.0.20171015/bin
patching script interpreter paths in /nix/store/0b5bvyqz6hy754c5cm2imsln59kjprvd-proot-5.1.0.20171015
checking for references to /build in /nix/store/0b5bvyqz6hy754c5cm2imsln59kjprvd-proot-5.1.0.20171015...
/nix/store/rshs7r6pm9d78mxfvx7b44pahq9zkcls-c-ares-1.14.0
/nix/store/h73hfs1kkr9r8zr8rgjzivdc2ff0azdq-libxml2-2.9.8-bin
/nix/store/iirmif7qgp7pgbv80z5x1sj3hbay893d-ncurses-6.1
/nix/store/qvsgzbf8h2b5fxrg80mzlvj59i28xlny-openssl-1.0.2p-bin
/nix/store/0ingn8cwwnl84i374hcl6nafsm2c5m2p-perl-5.28.0
/nix/store/0b5bvyqz6hy754c5cm2imsln59kjprvd-proot-5.1.0.20171015
/nix/store/di0d17idlw6fxf3qwpfyxciw94ywi664-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=87c2a285-87cc-46b3-8432-233242f676de-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=8920b357-2ff4-4a39-bbfc-26187836096c)

Attempted: c-ares, libxml2, ncurses, openssl, perl, proot, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/6s14vh4mlxvyp2kmd19rw13xsmbz4bff-proot-5.1.0/bin
patching script interpreter paths in /nix/store/6s14vh4mlxvyp2kmd19rw13xsmbz4bff-proot-5.1.0
checking for references to /build in /nix/store/6s14vh4mlxvyp2kmd19rw13xsmbz4bff-proot-5.1.0...
/nix/store/6mkaknm4k59rm54rr6gyi46m0gr17fpk-c-ares-1.14.0
/nix/store/sfvr5ib3yqyxj2vzr832in3dsihqsriy-libxml2-2.9.8-bin
/nix/store/wydzkhfm25cgbr1nfq28nqvgk0mk7518-ncurses-6.1
/nix/store/pz3zwqqdbdmnnbz0xqifnjdni0xh0bsi-openssl-1.0.2p-bin
/nix/store/bzgcpj9fijmlxgjaa72fjys7m3ycijy8-perl-5.28.0
/nix/store/6s14vh4mlxvyp2kmd19rw13xsmbz4bff-proot-5.1.0
/nix/store/x0aj2fsd3hskn1fckk3w4b8bl6mx8m1s-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=03d72983-a991-4791-8533-0ced0600af67-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=a0267f1b-bc35-4983-9a7f-0e5e1b0bb3e3)

Attempted: c-ares, ncurses, openssl, perl, proot, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/0b5bvyqz6hy754c5cm2imsln59kjprvd-proot-5.1.0.20171015/bin
patching script interpreter paths in /nix/store/0b5bvyqz6hy754c5cm2imsln59kjprvd-proot-5.1.0.20171015
checking for references to /build in /nix/store/0b5bvyqz6hy754c5cm2imsln59kjprvd-proot-5.1.0.20171015...
/nix/store/rshs7r6pm9d78mxfvx7b44pahq9zkcls-c-ares-1.14.0
/nix/store/iirmif7qgp7pgbv80z5x1sj3hbay893d-ncurses-6.1
/nix/store/qvsgzbf8h2b5fxrg80mzlvj59i28xlny-openssl-1.0.2p-bin
/nix/store/0ingn8cwwnl84i374hcl6nafsm2c5m2p-perl-5.28.0
/nix/store/0b5bvyqz6hy754c5cm2imsln59kjprvd-proot-5.1.0.20171015
/nix/store/di0d17idlw6fxf3qwpfyxciw94ywi664-zlib-1.2.11
```
</p></details>

Ericson2314 commented on pkgs/stdenv/cross/default.nix:
```diff
@@ -26,19 +26,24 @@ in lib.init bootStages ++ [
+      vanillaPackages.stdenv.override {
+        targetPlatform = if crossSystem == null
+                         then localSystem
+                         else crossSystem;
```
I think would be better to in `pkgs/top-level/default.nix` default `crossSystem` to be `localSystem`. Then all these can stay the same.

Ericson2314 commented on pkgs/stdenv/cross/default.nix:
```diff
@@ -48,14 +51,17 @@ in lib.init bootStages ++ [
              [ buildPackages.patchelf buildPackages.paxctl ]
+        ++ lib.optionals
+             (hostPlatform.isDarwin && buildPlatform.isDarwin)
+             [ buildPackages.clang ]
```
what's this?

matthewbauer replied:
We are using gcc for all cross stuff. But, gcc alone can't build binaries for some reason. It needs to call clang at some point.

matthewbauer replied:
This is the error I get without it:

```
/nix/store/...-gcc-wrapper-7.3.0/bin/as: assembler (clang) not installed
```

GrahamcOfBorg commented:
<!--REQUEST_ID=0a06529e-d60e-44db-aeec-350ada98b5e4-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=ba5082c3-622e-4e9d-b8e1-bf50d1ed0ef8)

Attempted: c-ares, ncurses, openssl, perl, stdenv, zlib

The following builds were skipped because they don't evaluate on x86_64-darwin: proot

<details><summary>Partial log (click to expand)</summary><p>

```
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/pnf2j3642cx94bsz21wyz194mv76zivn-c-ares-1.15.0
/nix/store/ja4av83fr8mnqxamkf7dwy4rw51k0inh-ncurses-6.1
/nix/store/9vidwdsdm5iawn5law05xy412x3qdbl8-openssl-1.0.2p-bin
/nix/store/maxij11md81wwwc9nh9c7nz40b7m0n1w-perl-5.28.0
/nix/store/ljh3mi6z5xdqz92vp99pilq6z242v3na-stdenv-darwin
/nix/store/xa206623g1r8z56x9qaf2kqzvk5qvrb3-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=0a06529e-d60e-44db-aeec-350ada98b5e4-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=e367822c-6434-4fc4-8123-de7c87258271)

Attempted: c-ares, ncurses, openssl, perl, proot, stdenv, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/sa95f5njr6rglrk6y2sl06h6vpmyqjgl-proot-5.1.0.20171015/bin
patching script interpreter paths in /nix/store/sa95f5njr6rglrk6y2sl06h6vpmyqjgl-proot-5.1.0.20171015
checking for references to /build in /nix/store/sa95f5njr6rglrk6y2sl06h6vpmyqjgl-proot-5.1.0.20171015...
/nix/store/8r7ijp8iqkh6j2f6cw7kazjbsa7a256y-c-ares-1.15.0
/nix/store/9xhs46vn458874r7im4533d27dhdy6m0-ncurses-6.1
/nix/store/wprnsc7ifyhdk2iij43dblwf8xg73k2q-openssl-1.0.2p-bin
/nix/store/svqh993szsgi152vywx5ybyh1k22l5j1-perl-5.28.0
/nix/store/sa95f5njr6rglrk6y2sl06h6vpmyqjgl-proot-5.1.0.20171015
/nix/store/f2cjn7937v93q7kq1nxy42z7h2navq0m-stdenv-linux
/nix/store/q9kcy4r6fw5n2qmnpg4daf8injp7d64g-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=0a06529e-d60e-44db-aeec-350ada98b5e4-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=e27ce87a-9d3e-4200-9c57-cde8e9380f5d)

Attempted: c-ares, ncurses, openssl, perl, proot, stdenv, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/z9c3dcwn2jkvqddz33yj5kxyydcighsw-proot-5.1.0/bin
patching script interpreter paths in /nix/store/z9c3dcwn2jkvqddz33yj5kxyydcighsw-proot-5.1.0
checking for references to /build in /nix/store/z9c3dcwn2jkvqddz33yj5kxyydcighsw-proot-5.1.0...
/nix/store/d18x3g8zgfrklv7icklspf05h4m8714x-c-ares-1.15.0
/nix/store/cz2gp7aicnwkmid3ixab0kyyb7sas8dd-ncurses-6.1
/nix/store/43yr9rihhfccpqzgj65l82mxkf9aad53-openssl-1.0.2p-bin
/nix/store/nv50smxc219sc49cwgvp0fi9w172y1b4-perl-5.28.0
/nix/store/z9c3dcwn2jkvqddz33yj5kxyydcighsw-proot-5.1.0
/nix/store/vxg1lrr6vpxqhvg402z099wvkb3qbby2-stdenv-linux
/nix/store/0s979pchvsfaz0svsb9cyzyxyckhl8ij-zlib-1.2.11
```
</p></details>

Ericson2314 commented on pkgs/stdenv/custom/default.nix:
```diff
@@ -2,7 +2,7 @@
 }:
 
-assert crossSystem == null;
+assert crossSystem == localSystem;
```
We're getting closer to changing these back to `buildPlatform` and `hostPlatform` :). Good discovery!

matthewbauer commented:
@grahamcofborg build pkgsStatic.hello

GrahamcOfBorg commented:
<!--REQUEST_ID=6d99a6c1-bdc0-495b-b8d0-ce4192e70826-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=cea76e5c-0ab0-45fe-ba16-487309f0b82c)

Attempted: c-ares, ncurses, openssl, perl, proot, stdenv, zlib

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/sa95f5njr6rglrk6y2sl06h6vpmyqjgl-proot-5.1.0.20171015/bin
patching script interpreter paths in /nix/store/sa95f5njr6rglrk6y2sl06h6vpmyqjgl-proot-5.1.0.20171015
checking for references to /build in /nix/store/sa95f5njr6rglrk6y2sl06h6vpmyqjgl-proot-5.1.0.20171015...
/nix/store/8r7ijp8iqkh6j2f6cw7kazjbsa7a256y-c-ares-1.15.0
/nix/store/9xhs46vn458874r7im4533d27dhdy6m0-ncurses-6.1
/nix/store/wprnsc7ifyhdk2iij43dblwf8xg73k2q-openssl-1.0.2p-bin
/nix/store/svqh993szsgi152vywx5ybyh1k22l5j1-perl-5.28.0
/nix/store/sa95f5njr6rglrk6y2sl06h6vpmyqjgl-proot-5.1.0.20171015
/nix/store/f2cjn7937v93q7kq1nxy42z7h2navq0m-stdenv-linux
/nix/store/q9kcy4r6fw5n2qmnpg4daf8injp7d64g-zlib-1.2.11
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=81625ac1-25f7-4b06-9ff7-e517e504dc22-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=1c9a9666-b3f4-403d-a7f8-980dc30aac03)

Attempted: pkgsStatic.hello

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/r1vx8ig6iq8amvx9kpzgrwwsywv87p6m-hello-2.10-x86_64-unknown-linux-musl
shrinking /nix/store/r1vx8ig6iq8amvx9kpzgrwwsywv87p6m-hello-2.10-x86_64-unknown-linux-musl/bin/hello
cannot find section .dynamic
gzipping man pages under /nix/store/r1vx8ig6iq8amvx9kpzgrwwsywv87p6m-hello-2.10-x86_64-unknown-linux-musl/share/man/
x86_64-unknown-linux-musl-strip is /nix/store/bm50cndgf2wh9v9vn4z9b5i73ydj5aan-x86_64-unknown-linux-musl-binutils-2.30/bin/x86_64-unknown-linux-musl-strip
stripping (with command x86_64-unknown-linux-musl-strip and flags -S) in /nix/store/r1vx8ig6iq8amvx9kpzgrwwsywv87p6m-hello-2.10-x86_64-unknown-linux-musl/lib  /nix/store/r1vx8ig6iq8amvx9kpzgrwwsywv87p6m-hello-2.10-x86_64-unknown-linux-musl/bin
patching script interpreter paths in /nix/store/r1vx8ig6iq8amvx9kpzgrwwsywv87p6m-hello-2.10-x86_64-unknown-linux-musl
checking for references to /build in /nix/store/r1vx8ig6iq8amvx9kpzgrwwsywv87p6m-hello-2.10-x86_64-unknown-linux-musl...
cannot find section .dynamic
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=81625ac1-25f7-4b06-9ff7-e517e504dc22-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48803&attempt_id=6d0164e2-f051-496e-9621-e65e7b341dca)

Attempted: pkgsStatic.hello

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/jgg01p587s3713mlqq3z016visnmsa0a-hello-2.10-aarch64-unknown-linux-musl
shrinking /nix/store/jgg01p587s3713mlqq3z016visnmsa0a-hello-2.10-aarch64-unknown-linux-musl/bin/hello
cannot find section .dynamic
gzipping man pages under /nix/store/jgg01p587s3713mlqq3z016visnmsa0a-hello-2.10-aarch64-unknown-linux-musl/share/man/
aarch64-unknown-linux-musl-strip is /nix/store/4vsgprn279kdpncvfdc4bgzaqc1k3csj-aarch64-unknown-linux-musl-binutils-2.30/bin/aarch64-unknown-linux-musl-strip
stripping (with command aarch64-unknown-linux-musl-strip and flags -S) in /nix/store/jgg01p587s3713mlqq3z016visnmsa0a-hello-2.10-aarch64-unknown-linux-musl/lib  /nix/store/jgg01p587s3713mlqq3z016visnmsa0a-hello-2.10-aarch64-unknown-linux-musl/bin
patching script interpreter paths in /nix/store/jgg01p587s3713mlqq3z016visnmsa0a-hello-2.10-aarch64-unknown-linux-musl
checking for references to /build in /nix/store/jgg01p587s3713mlqq3z016visnmsa0a-hello-2.10-aarch64-unknown-linux-musl...
cannot find section .dynamic
```
</p></details>

matthewbauer commented:
@GrahamcOfBorg build pkgsStatic.haskellPackages.aeson

matthewbauer committed:
stdenv: implement crossOverlays

crossOverlays only apply to the packages being built, not the build
packages. It is useful when you don’t care what is used to build your
packages, just what is being built. The idea relies heavily on the
cross compiling infrastructure. Using this implies that we need to
create a cross stdenv.

matthewbauer committed:
treewide: remove static packages

Nixpkgs is meant to link everything dynamically. We don’t want to
expose static packages at the top level. If some package needs
statically built binaries, it should use a custom override.

matthewbauer committed:
stdenv/adapters.nix: fixup makeStaticBinaries

- makeStaticBinaries don’t work on Darwin (no stable ABI!)
- Need to make sure NIX_CFLAGS_LINK appends
- isStatic is not used anymore

matthewbauer committed:
darwin/libiconv: fix static/shared logic

The conditional was incorrect - postInstall script should only hsppen
when enableShared = true.

matthewbauer committed:
zlib: clean up static/shared distincion

This is kind of a mess, but basically:

- static=true, shared=true means to build statically but move it to
  the static output
- static=true, shared=false means to build statically and leave it in
  the main output
- static=false, shared=true means to not build static at all

Confusingly, the old default was static=true, shared=true even though
static=false? Still can’t figure out what was meant by that.

matthewbauer committed:
ncurses: make shared linking optinoal

when enableShared = false, we set --without-shared flag.

matthewbauer committed:
top-level/stage.nix: add static overlay

Adds the static overlay that can be used to build Nixpkgs statically.
Can be used like:

  nix build pkgsStatic.hello

Not all packages build, as some rely on dynamic linking.

roberth commented on pkgs/top-level/default.nix:
```diff
@@ -22,16 +22,18 @@
   overlays ? []
 
+, # List of overlays to apply to target packages only.
+  crossOverlays ? []
```
This looks like a great improvement. As @Ericson2314 mentioned, this brings us much closer to `build` and `host` terminology. So what about changing this to `hostOverlays`, because then it can be applied to non-cross as well.

Ericson2314 replied:
With non cross, the build and host packages are the same, so it's not possible to just override in one case. `crossOverrides` actually will turn on cross if no `crossSystem` (to be renamed :)) was given to make things work.

matthewbauer replied:
One reason that crossOverlays makes sense is that you will be using the "cross" stdenv here. This could be a little surprising, for instance, if you use crossOverlays on macOS, you will get CC=gcc instead of CC=clang.

roberth replied:
@matthewbauer so that's the `host` or `run` `CC`, right? So why not `hostOverlays` or `runOverlays`? I'd say "cross" is a property of the entire package set: was it cross-compiled? yes or no, boolean. When describing which architecture some variable applies to, it should be `build` or `host`/`run`, because that is precise.

Now I might be entirely wrong about this, but this, `crossOverlays`, seems very similar to the way in which I accidentally wrote `pkgs.extend` to be [broken](https://github.com/NixOS/nixpkgs/commit/889a5c35127e4e6b2c722611dcaa9c802337446b#commitcomment-31560635).

I'm trying to do something at the edge of my understanding of Nixpkgs internals here on a limited time budget. My initial goal was to achieve greater consistency, mainly between the Nixpkgs function and NixOS, so that it will be as simple as possible to understand the interface and work with it. I hope that a new-ish perspective is useful; if not, I should focus on other things.

matthewbauer replied:
@robeth That looks like okay logic. We definitely want to unify the build, host, target & local, cross, eventually. But I still am thinking crossOverlays is good here because it mirrors crossSystem. Right now, this is how I see it-

- {build,host,target}Platform - used internally in Nixpkgs as the "immediate" values of build, host, target. These are given to you, not set directly. Sliding windows algorithm handles the details...
- {local,cross}System - external arguments of Nixpkgs describing **where** you are and where you want to go. This is what you want to set to get from point a to point b.

If cross-compiling is a graph, then each node is a set of packages. We activate a kind of greedy pathfinding algorithm in moving from localSystem to crossSystem. Outside of Nixpkgs, we shouldn't care about the exact path that is taken, just that we end up with a package built by "localSystem", for "crossSystem". In the same way, when you set "crossOverlays", you are saying to give me a package set that is statically built, but using my localSystem to avoid rebooting everything statically.

dylex commented:
Since this change, I get an error about passing an unknown crossOverlays argument to stdenv when using a custom stdenv.  2ee9522a6c76e28949b6237928d2cc74fb9edfc2 seems to fix it fine as far as I can tell, similar to what was done to the other stdenvs.  Does this seem right?

matthewbauer commented:
Thanks! I forgot to do some of the extra stdenv's. We could also just use ```...``` because we aren't using crossOverlays. Merged in 04beae0979007f5a5b99e64fc945211824a91572.

matthewbauer commented on pkgs/stdenv/adapters.nix:
```diff
@@ -32,13 +32,15 @@ rec {
+      if stdenv.hostPlatform.isDarwin
+      then throw "Cannot build fully static binaries on Darwin/macOS"
+      else stdenv.mkDerivation (args // {
+        NIX_CFLAGS_LINK = toString (args.NIX_CFLAGS_LINK or "") + "-static";
```
Yeah probably. The original version left out the space so I just left it as is

jb55 commented:
I've been playing with pkgsStatic today, I tried to build something which eventually required `meson` and then I hit this:

$ nix build nixpkgs.pkgsStatic.surf --show-trace

    error: while evaluating the attribute 'buildInputs' of the derivation 'surf-2.0-x86_64-unknown-linux-musl' at /home/jb55/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:187:11:
    while evaluating the attribute 'buildInputs' of the derivation 'webkitgtk-2.22.6-x86_64-unknown-linux-musl' at /home/jb55/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:187:11:
    while evaluating the attribute 'nativeBuildInputs' of the derivation 'gst-plugins-base-1.14.4-x86_64-unknown-linux-musl' at /home/jb55/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:187:11:
    while evaluating the attribute 'out.outPath' at /home/jb55/nixpkgs/lib/customisation.nix:142:13:
    while evaluating the attribute 'crossFile' of the derivation 'meson-0.49.0-x86_64-unknown-linux-musl' at /home/jb55/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:187:11:
    while evaluating the attribute 'text' of the derivation 'cross-file.conf' at /home/jb55/nixpkgs/pkgs/build-support/trivial-builders.nix:7:14:
    while evaluating the attribute 'stdenv.targetPlatform.parsed.kernel.name' at /home/jb55/nixpkgs/pkgs/stdenv/booter.nix:111:5:
    attribute 'targetPlatform' missing, at /home/jb55/nixpkgs/pkgs/development/tools/build-managers/meson/default.nix:70:17

it seems to be complaining about this section in meson's derivation:

    [host_machine]
    system = '${targetPackages.stdenv.targetPlatform.parsed.kernel.name}'
    cpu_family = '${targetPackages.stdenv.targetPlatform.parsed.cpu.family}'
    cpu = '${targetPackages.stdenv.targetPlatform.parsed.cpu.name}'
    endian = ${if targetPackages.stdenv.targetPlatform.isLittleEndian then "'little'" else "'big'"}

For some reason stdenv.targetPlatform is missing.

Any ideas? I'm not too familiar with the cross compilation
infrastructure to fix this.

Cheers,
Will

matthewbauer commented:
Yeah meson looks to be broken right now.
